mathics/builtin/numbers/algebra.py

```diff
@@ -1351,9 +1351,13 @@ def apply(self, expr, form, h, evaluation):
             return Expression(SymbolDirectedInfinity, Integer(-1))
 
         if not form.has_form("List", None):
```

Comment:
We could probably write a specialized version of `has_form("List", None) which I believe comes up a bit. Here, all you need to do is check that Head is `SymbolList`.  

We should benchmark to see if it is worth it though.

Reply:
Maybe even better would be to define List as a BuiltinExpression, which has at least two advantages:
* it would allow to check if is a List by using the standard way `isinstance(x, List)`
* Provides a way to have a simpler evaluation method: ``List`` is a Locked symbol, so all its attributes are always the same.

Reply:
I wholeheartedly agree!  Something like this has been in the back of my mind for a while.

The other aspect here that I would like to see in this object is an optional representation as a Python or NumPy list (when that is possible) so that we reduce conversion for those operations that are done in SymPy, numpy, mpmath, or Python. 

But that can come later.

As I have asked before, in Mathematica can you redefine List in any way? If not, the parser could form List literals directly.

Reply:
> I wholeheartedly agree! Something like this has been in the back of my mind for a while.
> 
> The other aspect here that I would like to see in this object is an optional representation as a Python or NumPy list (when that is possible) so that we reduce conversion for those operations that are done in SymPy, numpy, mpmath, or Python.
> 

Yep: I guess it could be helpful in some places like in linear algebra or graphics packages. 

> But that can come later.
> 
> As I have asked before, in Mathematica can you redefine List in any way? If not, the parser could form List literals directly.

In WL, ``List`` is ``Locked`` and ``Protected``, so you can only modify its behavior by defining upvalues
(something like ``List[MySymbol, a,b]^:=g[a,b]``) (because the rule is attached to ``MySymbol`` and not to ``List``.

Reply:
> In WL, `List` is `Locked` and `Protected`, so you can only modify its behavior by defining upvalues (something like `List[MySymbol, a,b]^:=g[a,b]`) (because the rule is attached to `MySymbol` and not to `List`.


Ok - I am not sure I understand this fully. But this means that whenever the parser sees "List" it is a Mathics List and therefore it could go directly into this ListExpression object, right? 

BTW when ListExpression exists, then Range would return that rather than an Expression.

ListExpression with the property elements_fully_evaluated means we have a list literal and that might be an important property to note in of itself.

Reply:
> 
> Ok - I am not sure I understand this fully. But this means that whenever the parser sees "List" it is a Mathics List and therefore it could go directly into this ListExpression object, right?

Yep

> 
> BTW when ListExpression exists, then Range would return that rather than an Expression.
> 

Exactly

> ListExpression with the property elements_fully_evaluated means we have a list literal and that might be an important property to note in of itself.

yes, because then, in construction like `F[{....}]`, the `evaluation` method of the expression does not need to try to evaluate the list.

Reply:
> yes, because then, in construction like F[{....}], the evaluation method of the expression does not need to try to evaluate the list.

Actually, that is currently how expression-interface-refactor works [right now]( https://github.com/Mathics3/mathics-core/blob/expression-interface-refactor/mathics/core/expression.py#L1262-L1270).